src/typeahead/typeahead.ts

```diff
@@ -0,0 +1,168 @@
+    '(document:click)': '_closePopup()',
+    '(input)': 'onChange($event.target.value)',
+    '(keydown)': '_handleKeyDown($event)'
+  },
+  providers: [NGB_TYPEAHEAD_VALUE_ACCESSOR]
+})
+export class NgbTypeahead implements OnInit,
+    ControlValueAccessor {
+  private _onChangeNoEmit;
+  private _popupService: PopupService<NgbTypeaheadWindow>;
+  private _positioning = new Positioning();
+  private _valueChanges = new EventEmitter<string>();
+  private _windowRef: ComponentRef<NgbTypeaheadWindow>;
+
+  /**
+   * A function to convert a given value into string to display in the input field
+   */
+  @Input() inputFormatter: (value) => string;
+
+  /**
+   * A function to transform the provided observable text into the array of results
+   */
+  @Input() ngbTypeahead: (text: Observable<string>) => Observable<any[]>;
+
+  /**
+   * A function to format a given result before display. This function should return a formatted string without any
+   * HTML markup.
+   */
+  @Input() resultFormatter: (value) => string;
+
+  /**
+   * A template to display a matching result.
+   */
+  @Input() resultTemplate: TemplateRef<ResultTplCtx>;
+
+  onChange = (value) => {
+    this._onChangeNoEmit(value);
+    this._valueChanges.emit(value);
+  };
+
+  onTouched = () => {};
+
+  constructor(
+      private _elementRef: ElementRef, private _viewContainerRef: ViewContainerRef, private _renderer: Renderer,
+      private _injector: Injector, componentFactoryResolver: ComponentFactoryResolver) {
+    this._popupService = new PopupService<NgbTypeaheadWindow>(
+        NgbTypeaheadWindow, _injector, _viewContainerRef, _renderer, componentFactoryResolver);
+    this._onChangeNoEmit = (_: any) => {};
+  }
+
+  ngOnInit() {
+    this._valueChanges.asObservable().let (this.ngbTypeahead).subscribe((results) => {
+      if (!results || results.length === 0) {
+        this._closePopup();
+      } else {
+        this._openPopup();
+        this._windowRef.instance.results = results;
+        this._windowRef.instance.term = this._elementRef.nativeElement.value;
+        if (this.resultFormatter) {
+          this._windowRef.instance.formatter = this.resultFormatter;
+        }
+        if (this.resultTemplate) {
+          this._windowRef.instance.resultTemplate = this.resultTemplate;
+        }
+      }
+    });
+  }
+
+  registerOnChange(fn: (value: any) => any): void { this._onChangeNoEmit = fn; }
+
+  registerOnTouched(fn: () => any): void { this.onTouched = fn; }
+
+  writeValue(value) {
+    const formattedValue = value && this.inputFormatter ? this.inputFormatter(value) : toString(value);
+    this._renderer.setElementProperty(this._elementRef.nativeElement, 'value', formattedValue);
+  }
+
+  private _closePopup() {
+    this._popupService.close();
+    this._windowRef = null;
+  }
+
+  private _handleKeyDown(event: KeyboardEvent) {
+    if (!this._windowRef) {
+      return;
+    }
+
+    if (['ArrowDown', 'ArrowUp', 'Enter', 'Escape'].indexOf(event.key) > -1) {
```

Comment:
Are you sure that this is going to work fine cross-browser? This is what we are doing in the angular-ui/bootstrap: https://github.com/angular-ui/bootstrap/blob/c61d16a9330c18105d074ffd23f14db06689809b/src/typeahead/typeahead.js#L399-L431